kore/src/Kore/Step/Axiom/Matcher.hs

```diff
@@ -9,93 +9,494 @@ Stability   : experimental
+            field @"queued" .= queued'
+            return (Just next)
+        _ ->
+            return Nothing
+
+{- | Push a new constraint onto the matching queue.
+ -}
+push
+    :: MonadState (MatcherState variable) matcher
+    => Pair (TermLike variable)
+    -> matcher ()
+push pair = field @"queued" %= (pair :<|)
+
+{- | Defer a constraint until more information is available.
+
+The constraint will be retried after the next substitution which affects it.
+
+ -}
+defer
+    :: MonadState (MatcherState variable) matcher
+    => Pair (TermLike variable)
+    -> matcher ()
+defer pair = field @"deferred" %= (:|> pair)
+
+{- | Record a substitution in the matching solution.
+
+The substitution is applied to the remaining constraints and to the partial
+matching solution (so that it is always normalized). @substitute@ ensures that
+the variable does not occur on the right-hand side of the substitution and adds
+a constraint that the right-hand side is defined (if necessary).
+
+ -}
+substitute
+    :: (MatchingVariable variable, MonadUnify unifier)
+    => variable
+    -> TermLike variable
+    -> MaybeT (MatcherT variable unifier) ()
+substitute variable termLike = do
+    -- Ensure that the variable does not occur free in the TermLike.
+    occursCheck variable termLike
+    -- Ensure that the TermLike is defined.
+    definedTerm termLike
+    -- Record the substitution.
+    field @"substitution" <>= Map.singleton variable termLike
+
+    -- Isolate the deferred pairs which depend on the variable.
+    -- After substitution, the dependent pairs go to the front of the queue.
+    MatcherState { deferred } <- Monad.State.get
+    let (indep, dep) = Seq.partition isIndependent deferred
+    field @"deferred" .= indep
+
+    -- Push the dependent deferred pairs to the front of the queue.
+    Foldable.traverse_ push dep
+    -- Apply the substitution to the queued pairs.
+    field @"queued" . Lens.mapped %= substitute2
+
+    -- Apply the substitution to the accumulated matching solution.
+    field @"substitution" . Lens.mapped %= substitute1
+    field @"predicate" . Lens.mapped %= Syntax.Predicate.substitute subst
+
+    return ()
+  where
+    isIndependent = not . any (hasFreeVariable variable)
+    subst = Map.singleton variable termLike
+    substitute2 = fmap substitute1
+    substitute1 = renormalize . TermLike.substitute subst
+
+{- | Renormalize builtin types after substitution.
+ -}
+renormalize
+    :: MatchingVariable variable
+    => TermLike variable -> TermLike variable
+renormalize termLike =
+    case termLike of
+        BuiltinMap_ internalMap ->
+            Lens.traverseOf (field @"builtinAcChild") Ac.renormalize internalMap
+            & maybe bottom' mkBuiltinMap
+        BuiltinSet_ internalSet ->
+            Lens.traverseOf (field @"builtinAcChild") Ac.renormalize internalSet
+            & maybe bottom' mkBuiltinSet
+        _ -> termLike
+  where
+    bottom' = mkBottom (termLikeSort termLike)
+
+occursCheck
```

Comment:
Would it be better to put this in `Kore.Unification.SubstitutionNormalization` near `nonSimplifiableSortResult`?

Reply:
Well, now I'm not so sure about the implementation. In unification, we throw an error for `x = f(x)` substitutions (`f` not a constructor) because we could, possibly, handle those. For matching, on the other hand, we probably want to say intentionally that `x` does not match `f(x)`; in that case this would be
```.hs
occursCheck variable termLike = (Monad.guard . not) (hasFreeVariable variable termLike)
```
in which case it should probably not be used outside the matcher.